### ISSUE TITLE
Saga Concurrent Updates

### ISSUE THREAD

**Issue by MrMDavidson** [rebus-org/Rebus#341]
We're having an issue with Rebus where multiple workers are attempting to work on a saga and they're hitting a `ConcurrencyException`. We're using the SQL transport and saga storage. A reproduction case is included (just update `connectionString` before building) and tested against the current NuGet package (0.98.5). Once the app spins up it'll you'll be presented with 3 options. Hit "1" to add an additional worker. "2" to remove a worker. And "3" to start the saga. To begin with add a single worker and then press 3. It'll queue a `StartSagaCommand` command that initiates a new `CountingSaga`. This simple saga sends 100 `IncrementCounterCommand`s which will be handled by the same saga. When a `IncrementCounterCommand` a counter is incremented in the saga data and if we've received the 100 messages we mark the saga as complete and print a message. You'll see lots of `ConcurrencyExceptions` thrown. If you add another worker (Hit 1) or fire up another copy of this app you'll get even more. To the point where the saga probably won't even finish.

(Side note: If I set `SetMaxParallelism()` to a lower value I have a better chance of success. But this still doesn't help if I have multiple workers. What's the purpose of this? What is a sane value? Here I've set it to 10)

From my understanding what's happening is that Worker Thread A is pulling a message, `LoadSagaDataStep` is executing, seeing it's in a saga, loading the saga data at version N, awaiting the work for that worker, and then updating the saga. In the `SqlServerSagaStorage` `Update(...)` method an `UPDATE ... WHERE ... [Revision] = $CurrentRevision$` is performed and, if no rows are found, it throws a `ConcurrencyException`. As more worker threads are operating the chance of the revision being incremented between loading the saga and updating it increases. 

So, two things...
1. Am I doing something completely wrong here? Admittedly this example doesn't do a great deal of work so multiple messages processing at about the same time seems very likely. But it's hard to get buy in from people if this causes failure.
2. I'm happy to work on a fix. My plan was to introduce a `ISagaLocker` interface. `LoadSagaDataStep` would then find-and-lock the saga with a timeout. If the saga couldn't be locked within the timeout I'd propose deferring the message and retrying later so that the worker thread isn't tied up. (An example `ISagaLocker` implementation might be an `SqlServerApplicationLockSagaLocker` which would use SQL app level locks - https://msdn.microsoft.com/en-us/library/ms189823.aspx). Once the other items in the process pipeline have executed the `LoadSagaDataStep` would (in a finally block, obviously!) release the saga lock. This would effectively make each saga "single-threaded". By default (so behaviour is not altered) I can do a NoOpSagaLocker that did nothing. Any thoughts on this?

``` C#
using System;
using System.Threading.Tasks;
using Rebus.Activation;
using Rebus.Bus;
using Rebus.Config;
using Rebus.Extensions;
using Rebus.Handlers;
using Rebus.Messages;
using Rebus.Persistence.SqlServer;
using Rebus.Retry.Simple;
using Rebus.Routing;
using Rebus.Sagas;
using Rebus.Transport.SqlServer;

namespace RebusStressTest {
    public class Program {
        public static string connectionString = @"";

        public class StartSagaCommand {
            public Guid Id { get; set; }
            public int NumberOfMessages { get; set; }
        }

        public class IncrementCounterCommand {
            public Guid Id { get; set; }
        }

        public class CountingSagaData : SagaData {
            public int NumberOfMessages { get; set; }
            public int ReceivedMessages { get; set; }
        }

        public class CountingSaga : Saga<CountingSagaData>,
            IAmInitiatedBy<StartSagaCommand>,
            IHandleMessages<IncrementCounterCommand> {
            private readonly IBus _bus;

            public CountingSaga(IBus bus) {
                _bus = bus;
            }

            /// <summary>
            /// This method must be implemented in order to configure correlation of incoming messages with existing saga data instances.
            ///             Use the injected <see cref="T:Rebus.Sagas.ICorrelationConfig`1"/> to set up the correlations, e.g. like so:
            /// <code>
            /// config.Correlate&lt;InitiatingMessage&gt;(m =&gt; m.OrderId, d =&gt; d.CorrelationId);
            ///             config.Correlate&lt;CorrelatedMessage&gt;(m =&gt; m.CorrelationId, d =&gt; d.CorrelationId);
            /// </code>
            /// </summary>
            protected override void CorrelateMessages(ICorrelationConfig<CountingSagaData> config) {
                config.Correlate<StartSagaCommand>(x => x.Id, x => x.Id);
                config.Correlate<IncrementCounterCommand>(x => x.Id, x => x.Id);
            }

            /// <summary>
            /// This method will be invoked with a message of type <typeparamref name="TMessage"/>
            /// </summary>
            public async Task Handle(StartSagaCommand message) {
                if (Data.NumberOfMessages != 0) {
                    return;
                }

                Data.Id = message.Id;
                Data.NumberOfMessages = message.NumberOfMessages;

                for (int i = 0; i < Data.NumberOfMessages; i++) {
                    await _bus.Send(new IncrementCounterCommand() { Id = Data.Id });
                }
            }

            /// <summary>
            /// This method will be invoked with a message of type <typeparamref name="TMessage"/>
            /// </summary>
            public Task Handle(IncrementCounterCommand message) {
                Data.ReceivedMessages++;

                return CheckAndMarkSagaForCompletion();
            }

            private Task CheckAndMarkSagaForCompletion() {
                if (Data.NumberOfMessages == Data.ReceivedMessages) {
                    Console.WriteLine("Saga {0} has completed!", Data.Id);
                    MarkAsComplete();
                }

                return Task.FromResult(0);
            }
        }

        private static void Main(string[] args) {
            string inputQueueName = "Test.Input";
            string errorQueueName = "Test.Error";

            using (BuiltinHandlerActivator handlerActivator = new BuiltinHandlerActivator()) {
                RebusConfigurer inputConfig = ConfigureBus(handlerActivator, connectionString, inputQueueName, errorQueueName, false);
                using (IBus bus = inputConfig.Start()) {
                    handlerActivator.Register<IHandleMessages<StartSagaCommand>>(() => new CountingSaga(bus));
                    handlerActivator.Register<IHandleMessages<IncrementCounterCommand>>(() => new CountingSaga(bus));

                    bus.Advanced.Topics.Subscribe(typeof (StartSagaCommand).GetSimpleAssemblyQualifiedName());
                    bus.Advanced.Topics.Subscribe(typeof (IncrementCounterCommand).GetSimpleAssemblyQualifiedName());

                    Console.WriteLine("Bus has started. Press any of the following keys");
                    Console.WriteLine("1 = Add Worker");
                    Console.WriteLine("2 = Remove Worker");
                    Console.WriteLine("3 = Start Saga");
                    Console.WriteLine("Q = Quit");

                    ConsoleKeyInfo input = Console.ReadKey();
                    while (input.KeyChar != 'Q') {
                        Console.WriteLine();
                        switch (input.KeyChar) {
                            case '1': {
                                bus.Advanced.Workers.SetNumberOfWorkers(bus.Advanced.Workers.Count + 1);
                                Console.WriteLine("Now running with {0} workers", bus.Advanced.Workers.Count);

                                break;
                            }
                            case '2': {
                                if (bus.Advanced.Workers.Count > 0) {
                                    bus.Advanced.Workers.SetNumberOfWorkers(bus.Advanced.Workers.Count - 1);
                                    Console.WriteLine("Now running with {0} workers", bus.Advanced.Workers.Count);
                                } else {
                                    Console.WriteLine("Already have 0 workers.");
                                }
                                break;
                            }
                            case '3': {
                                bus.Send(new StartSagaCommand() { Id = Guid.NewGuid(), NumberOfMessages = 100 })
                                    .GetAwaiter()
                                    .GetResult();
                                break;
                            }
                        }

                        input = Console.ReadKey();
                    }

                    Console.WriteLine("Quitting");
                }
            }

            Console.ReadKey();
        }


        private static RebusConfigurer ConfigureBus(IHandlerActivator activator, string sagaDatabaseConnectionString, string inputQueueName, string errorQueueName, bool transformToErrors = false) {
            RebusConfigurer config = Configure
                .With(activator)
                .Logging(x => x.ColoredConsole());

            config.Transport(x => x.Register(res => new SqlServerTransport(new DbConnectionProvider(sagaDatabaseConnectionString), "RebusMessage", inputQueueName)));
            config.Subscriptions(x => x.StoreInSqlServer(sagaDatabaseConnectionString, "RebusSubscription", automaticallyCreateTables: false));
            config.Sagas(x => x.StoreInSqlServer(sagaDatabaseConnectionString, "RebusSaga", "RebusSagaIndex", automaticallyCreateTables: false));
            config.Timeouts(x => x.StoreInSqlServer(sagaDatabaseConnectionString, "RebusTimeout", automaticallyCreateTables: false));
            config.Options(o => o.SimpleRetryStrategy(errorQueueName, maxDeliveryAttempts: 5));

            config.Routing(x => x.Register(res => new StaticRouter(inputQueueName)));

            config.Options(x => x.LogPipeline(verbose: true));

            config.Options(x => x.SetNumberOfWorkers(0));
            config.Options(x => x.SetMaxParallelism(10));

            return config;
        }
    }

    public class StaticRouter : IRouter {
        private readonly string _queue;

        public StaticRouter(string queue) {
            _queue = queue;
        }

        /// <summary>
        /// Called when sending messages
        /// </summary>
        public Task<string> GetDestinationAddress(Message message) {
            return Task.FromResult(_queue);
        }

        /// <summary>
        /// Called when subscribing to messages
        /// </summary>
        public Task<string> GetOwnerAddress(string topic) {
            return Task.FromResult(_queue);
        }
    }
}
```

**Comment by mookid8000** [comment]
I guess what you're experiencing is Rebus' severely improved capability of parallelizing operations... with all this `async`/`await` stuff going on, especially with the SQL Server driver that supports it so well.

In many scenarios, I'd go with 1 worker and allow maybe 10 parallel operations - but in your case that would still yield a lot of `ConcurrencyException`s because of the contention around each saga ID immediately after sending the `IncrementCounterCommand` to itself.

I think using SQL Server locking is the way to go, but I think a simpler approach would be to simply `SELECT (...) WITH (ROWLOCK, READPAST)` in [this query](https://github.com/rebus-org/Rebus/blob/master/Rebus/Persistence/SqlServer/SqlServerSagaStorage.cs#L178-L182).

This would of course require that the saga connection is held for the entire time while processing the message, but that is definitely doable.

Thoughts?

**Comment by MrMDavidson** [comment]
Yeah - in our scenario we have 1 worker thread per CPU and have the parallelism set to CPU count + 1. Then there's another worker thread working on the error queue. We're hosted in Azure so we autoscale based on load... so lowering thread count and upping parallel operations wouldn't work.

Holding the saga connection for the entire duration of the thread processing might be a quick win but I see two possible issues;
 1. If you have a long running handler the transaction that is associated with the saga connection may time out. Which would mean the work associated with the message processing is lost and needs to re-execute anyhow. With SQL app locks the locks can live for longer than the life of the connection. So the saga finder can close the connection, the handler can take minutes, and then the saga updater can release the lock.
 2. If you used `ROWLOCK, READPAST` hints then if Thread B tried to work on the same saga as Thread A it'd read past the row and re-create the state - unless you planned to handle it by deferring the message in that case? No sure how you'd tell if there was genuinely no state or not. If you meant a blocking lock here my concern would be that Thread B is just idle waiting for Thread A to finish working on that saga.

**Comment by mookid8000** [comment]
1. If you use the SQL transport, you have the same issue - it's holding on to the same connection for the entire time it takes until it can commit the queue transaction. Obviously not the best choice for handlers that take a long time to execute.

2. Good catch! Phew!

Application-level locks sound interesting - am I correct in my understanding that it would be possible to `sp_getapplock` on something like `string.Format("saga-{0}", sagaData.Id)` from within the `Find` method on `SqlServerSagaStorage`, and then just release it again once the queue transaction was disposed?

**Comment by MrMDavidson** [comment]
1. Ah, valid point. I might need to do some more testing around that for our specific scenario.

Yup - application level locks are a "named" resource that you define. So, yeah, you'd just name the lock based on the saga (`string.Format("rb-saga-{0}", saga.id)` as you say). One sticking point might be that if the process hosting Thread A completely dies... something needs to know to tell SQL Server to release the application lock. You *might* be able to get around that with the transactional semantics of app locks, but it's been a while since I've used them so I would have to do some more investigation. If not, then you're basically having to use some kind of lease (Eg. Extra column on the saga table that says you grabbed it at 9am) and clearing that lease if you haven't processed it after some sane value of time (which will vary per application).

**Comment by mookid8000** [comment]
Oh, I thought the timeout on the lock was for auto-releasing it - it's not, it's just how long to block before potentially getting the lock.

Well, in that case I'm thinking that your lease timestamp could do the trick, which would also be portable to the PostgreSQL saga storage... then we would also have fewer moving parts.

Updating an existing saga instance would then consist of this:

* find instance based on correlation property, just like now, but also select out the [locked] column and the value of `getdate()`
* check the lease timestamp - if it's `DbNull.Value` proceed, or if the time diff between it and `getdate()` is greater than the max lease age then proceed (but log a warning)
* issue an optimistic `update set [locked] = getdate() where [id] = @id and [revision] = @revision and [locked] is null`
* if results > 1 (we got the lock) proceed by returning the saga instance
* (process message)
* then remember to remove the lock timestamp again when the saga is updated or if the transaction is aborted.

Thoughts?

**Comment by MrMDavidson** [comment]
Sounds good. As the saga storage is per-database-provider we can use an `UPDATE... OUTPUT...` clause to get and retrieve the lock as one atomic operation on SQL Server without doing a second round trip.

I'd still like to look into supporting message deferral if the saga for the message is currently locked. Otherwise you're blocking a perfectly good worker thread.

**Comment by mookid8000** [comment]
....but if you `UPDATE....OUTPUT...` wouldn't you run into the same issue as `WITH (ROWLOCK, READPAST` that a new saga data instance could be created when it shouldn't?

**Comment by MrMDavidson** [comment]
I'll have to look at the exact query / structure involved. But something like the following should work;
```SQL
UPDATE
...
WHERE
  Id = @SagaId
AND
  LeasedBy IS NULL OR LeasedBy < @LeaseExpiredDateTime
OUTPUT
SELECT
  1 as LockAcquired,
  1 as StateExists,
  Data as SagaData
UNION ALL
SELECT
  0 as LockAcquired,
  1 as StateExists
  NULL AS SagaData,
...
WHERE Id = @SagaId
UNION
ALL
SELECT
  0 as LockAcquired,
  0 as StateExists,
  NULL as SagaData
```

Then when you inspect the results in C# you only look at the first row of the result set that we get back. The first row is that we acquired the lock on the saga and we state exists (obviously!), second row is that we can't lock the saga but it does exist (so we'd defer the message) and 3rd row is that the saga doesn't yet exist so we'd need to create it. If we lock the row we'd get all three rows, if we can't lock it but it exists we'd get the last two rows, and if it doesn't exist we'd only get the last row... but regardless we only inspect the first row of the recordset.

**Comment by mookid8000** [comment]
and regarding the blocking of the worker thread, a little bit of tolerance could maybe be built in by doing an `await Task.Delay(...)` between a couple of attempts at getting the lock before giving up and throwing the concurrency exception.....

**Comment by mookid8000** [comment]
wow, crazy SQL :grin:

**Comment by MrMDavidson** [comment]
I'm a twit - you can't actually do an `UPDATE.. OUTPUT..` in a `UNION`. That means it'll require two separate result sets in the query. Which isn't too bad.

Anyhow, I've started implementing this over at https://github.com/MrMDavidson/Rebus. I've implemented it for the Id-found-sagas so far (not ones found by a correlation property). Hit a bit of a roadblock though in regard to deferral of messages - I need an instance of an `IBus` in the saga step itself... or I need to duplicate the message deferral implementations. Any chance you could have a look at the repository and potentially implement such a thing? Basically at https://github.com/MrMDavidson/Rebus/blob/master/Rebus/Pipeline/Receive/LoadSagaDataStep.cs#L91 I'd like to chuck the message back on the queue.

Also, from looking at it I think there might still be the potential for race conditions / concurrency exceptions in the scenario where multiple messages want to start the same saga. Any thoughts around this?

**Comment by mookid8000** [comment]
This comment is going to be big, so I'll do it with proper sections and headlines and all :) first:

### Sorry

Hmm.... I'm sorry, but after having thought all day about how much work it actually takes to lock a specific saga data instance, I'm beginning to think that there's something smelly about the way we're trying to solve it. 

I also dislike the fact that the solution we come up with will most likely be very specific for SQL Server and will not readily be portable to e.g. MongoDB.

When I get this kind of bad feelings about things, I usually try to back out a bit, and try to re-think the whole situation....  What I'm thinking right now is this:

### The locking solution

Basically, the problem can be boiled down to this: We find yourself in a situation where we're trying to process messages in parallel, some of which must be serially processed because of the nature of the work that we're trying to do.

Now we're trying to solve the problem by introducing constraints into the parallel processing of the messages, effectively implementing exclusive access to certain parts of the work that we're supposed to do. 

My conclusion so far is that this way of solving the problem would turn out to be too cumbersome, and I fear that the solution would introduce more problems than it would solve. 

E.g. deferring messages will introduce latency into the processing of the messages, and we'll be absolutely sure that one single `ConcurrencyException` will result in reordered messages (not that you should rely on message ordering in general, but _some_ ordering can usually be depended on), the code to handle this looks to become complex, etc.

### Another solution?

Another way to solve the problem would be to just _not do the processing in parallel in the first place_! This way, there would be _no contention at all_.

So, let's go through a couple of ways that message processing can be serialized....

#### By configuration

The easy way would be to simple have one single endpoint with 

    .Options(o => o.SetNumberOfWorkers(1).SetMaxParallelism(1))

but then nothing would be processed in parallel, ever. 

A step up could be to have an endpoint with the above configuration do the saga processing, and then leave everything else to an endpoint that processes things in parallel.

#### By making race conditions less likely to happen

One situation where a race condition is pretty likely, is when the saga instance sends a bunch of messages to itself. While this is a totally OK thing to do, if it's more than a few messages, it's pretty likely that there will be contention because the messages will be received at the same time.

In this case, if it's not a problem, messages could be deferred and evenly distributed in time in the future.

If messages should be processed by the saga as fast as possible, the saga itself could hold the messages (i.e. representing a list of things to be done) which could then lead to invoking the saga by doing something like this:

    if (Data.ThingsToDo.Count > 0)
    {
        var thingToDo = Data.ThingsToDo.First();
        Data.ThingsToDo.Remove(thingToDo);
        await _bus.SendLocal(thingToDo);
    }
    else
    {
       await _bus.SendLocal(new AllDone {CorrelationId = Data.CorrelationId});
    }

after doing each thing.

### How big is the problem

Regardless of which solution you think would work, I also encourage you to consider whether the problem would actually be that big - i.e. how big is the risk that a message fails so many times because of `ConcurrencyException` that it gets moved to the error queue?

### If you're still sure you want to lock & defer...

...I will still be happy to help :) but we should not mess around in the existing `LoadSagaDataStep`, we should create a new `LockAndLoadSagaDataStep` (although ["load and lock"](https://en.wiktionary.org/wiki/lock_and_load) would probable be more appropriate :grin:)  that does the locking. In addition to this, I'd like the `ISagaStorage` interface to stay unchanged, so we would have to figure out some other way of communicating back from `Find` whether the lock was also successfully grabbed.

### So...

Sorry about the long silence followed by this immense load of text..... I haven't been near a computer all day, but as you can probably see, it's not because I haven't been thinking about it :grin: 

Let me know what you think!

**Comment by mookid8000** [comment]
PS: Regarding multiple messages that want to start the same saga - no way to avoid that, must be enforced by the data store's ability to keep a unique constraint on correlation properties.

**Comment by MrMDavidson** [comment]
No worries - didn't expect you to necessarily reply straight away.

### Re: by configuration

We definitely don't want to have completely single threaded access to all saga operations. This serialises not only SagaA's internal operations but makes it serialised with SagaB. Additionally in a cloud, autoscale, implementation you're then going to run into headaches about where that one thread runs, monitoring it, etc.

### Re: By making race conditions less likely to happen

To be honest, this is probably a good approach to the problem. The issue I have with it is that it requires specific implementation, of an infrastructural issue, at the application level. It's not so bad with one saga... but if you have to do it everytime?

Sidenote: One option here might be to introduce something like...
```C#
public class SagaBus {
  public void PublishMessages(IEnumerable<object> messages) {
    // Broadcast a BatchedMessageSagaCommand which basically does your Data.ThingstoDo() and then broadcasts some kind of finish event
  }
}
```

### Re: How big is the problem

Getting it to the point of moving to the poison queue in a real world application that performs real work is probably pretty slim. But it does result in wasted work. Interesting that [NServiceBus](http://docs.particular.net/nservicebus/sagas/concurrency) not only has this same problem but notes that it in high message scenarios you can permanently fail.

### Re: If you're still sure you want to lock & defer...

I think it is still an issue. I've modified my example so that the saga sends a bunch of commands. These commands are handled in a non-saga handler that sleeps for a random 250ms - 3000ms period and then defers a "Done" message for a random time between 1000ms and 10,000ms and we're still not completing the saga due to concurrency issues. (Although this may be due to deferral not being 100% accurate? I'm not sure)

I do think any implementation is going to be storage specific because how you'd lock something (optimally) on SQL Server is going to differ to how you lock it in Mongo, or the like. But I understand wanting to keep the scope of changes to a minimum. What I'd propose is;
 1. I'll revert my changes to `ISagaStorage`
 1. Introduce a new `ILockableSagaStorage` interface. This would have a `FindAndLockSaga()` and an `UpdateSagaAndReleaseLock()` method.
 1. Extend `SqlServerSagaStorage` to implement `ILocableSagaStorage` as well as `ISagaStorage`
 1. Refactor some of `LoadSagaStep` to be virtual
 1. Introduce a `LockAndLoadSagaStep` that derives from `LoadSagaStep`.
 1. Add a configuration extension to enable `LockAndLoadSagaStep` instead of `LoadSagaStep`. This would require that the application uses a `ILockableSagaStorage`

**Comment by mookid8000** [comment]
I've got yet another option that I suggest you think about before writing too much code.... how about an alternative retry strategy? NServiceBus has something they call "2nd level retries", which is a thing that kicks in once the ordinary, quick successive delivery attempts have failed too many times.

I think it's just a simple configurable behavior that determines how long a failed message will be deferred into the future.

It's a functionality that I've been thinking about adding to Rebus for a while, but old Rebus was not extensible enough that this could be done without too much headache - but I think new Rebus is extensible enough that a more advanced retry mechanism could be made, that could e.g. be configured to retry quickly a few times on `ConcurrencyException` followed by deferring the message a small random timespan.

I would prefer this solution over the locking strategy because a) it will work with all the existing saga storages and not require anything special from the infrastructure, and b) it can be used for many other things besides retrying more intelligently on `ConcurrencyException`s.

Only downside is that it would be an entirely new concept in Rebus and would probably require some kind of persistence mechanism too, and would maybe require a few attempts before it was good enough.

Actually, the more I think about it, I'm beginning to convince myself that this would be a good solution. What do you think?

**Comment by MrMDavidson** [comment]
See my comment over on #342 - we already have something akin to 2nd level retries. In the more-realistic-example that started this issue we are handling those fatal messages (to, eg, offset the number of expected messages) But even those can still permanently fail due to `ConcurrencyException`s.

It's definitely a good feature to have - and we will happily make use of it! If you want to take the `Error` queue stuff I mention over there and make it a first class citizen we currently spool up the second queue + worker ourselves... not a huge issue but it probably has wider application beyond us.

Unless you have any massive objections I'd still like to go ahead and implement what I outlined above. I might go ahead and have a crack at the plan I outlined above with the `ILockableSagaStorage` interface + implementation.

**Comment by mookid8000** [comment]
Sounds like you've made a pretty cool solution there, actually - I like the idea of having application-level handling of failure scenarios that the application knows how to handle.

Regarding the locking thing - could you start out by adding it in a separate project to keep things separate? Then we'll see how it turns out and we can think about adding it to the core later on...

**Comment by mookid8000** [comment]
Hello @MrMDavidson , how are you doing?

Is there any news on this?

**Comment by MrMDavidson** [comment]
Unfortunately not - some other priorities have popped up at work that have waylaid any progress. I'll try and make it a priority for next week.

**Comment by mookid8000** [comment]
Just had a thought - what if a saga had a method that could be overridden like this:

    public void override HandleConflict(TSagaData otherSagaData)
    {
        // merge stuff from otherSagaData into Data here...
    }

which would then allow for manually merging saga data when a conflict was detected...

Pros:

* no locking
* better scalability
* works with all existing saga persisters
* no locking

Cons:

* ???

**Comment by pruiz** [comment]
1) Would we be able to throw something from within such a method in case we
just want the existing behavior (ie. do nothing, let the message go back to
broker, and reprocess it later on).
2) May that method be part of an independent interface (ie.
ISagaCanHandleConflicts) as to make it opt-in?

**Comment by mookid8000** [comment]
> 1) Would we be able to throw something from within such a method in case we
just want the existing behavior (ie. do nothing, let the message go back to
broker, and reprocess it later on).

There should be no change in behaviour whatsoever, unless you explicitly opt in to handling conflicts like this.

> 2) May that method be part of an independent interface (ie.
ISagaCanHandleConflicts) as to make it opt-in?

I thought about that, but keeping it as an overridable method directly on the generic `Saga<TSagaData> where TSagaData : ISagaData` allows for using the generic type in the signature of the virtual method - e.g. `Saga<FunkySagaDizzle>` would have the

    public void override HandleConflict(FunkySagaDizzle otherSagaData) { ... }

signature. Don't think anything similar could be accomplished with an interface.

**Comment by pruiz** [comment]
Maybe I didnt understand what you mean, but:

```csharp
public interface ISagaCanHandleConflicts<TSagaData>
  where TSagaData : ISagaData
{
   void HandleConflict(FunkySagaDizzle otherSagaData);
}
```

:?

**Comment by mookid8000** [comment]
Oh, but your saga class def would then look like this:

    public class MySaga : Saga<MySagaData>, ISagaCanHandleConflicts<MySagaData> {...}

requiring that the saga data type be specified twice

**Comment by MrMDavidson** [comment]
I've had a chance to look into this a little bit today. I'm not convinced the `HandleConflict()` method is an appropriate solution for the concurrency issues I raised. It might be something worthwhile to introduce just as an "FYI: Your saga just threw `ConcurrencyException`!" although you'd need some way to indicate to Rebus what you want to do.

Anyway; unless I'm mistaken this isn't something that would happen until after the message handler has fired (as that's when the `ConcurrencyException` occurs). That means you've already performed work based on what you expected the state of the saga to be in whilst it could be different. This gets extra hairy when external systems are involved (Sure - the `IdempotentSaga` implementation would help here).

On the one hand I still think there's potential value in a "Single Access Saga" (based on the type / Id tuple) but I think it should definitely be something you opt into (ie. You tag the saga with an appropriate interface and set the queue up appropriately). But on the other... I could also very easily refactor this saga so that it was more of a Controller (eg. It received the start command. Sets up its state. Then sends a "Do Work" command. On response it sends the next "Do Work" command, etc ref: https://lostechies.com/jimmybogard/2013/03/21/saga-implementation-patterns-variations/). I initially balked at this because you lose the ability to scale but if you're making access to the saga id / type tuple single access you've lost that scaling anyway. 

I'll mull over the options for implementation over the weekend... but I'm starting to think it might be best to just tag this one a "will not fix".

**Comment by pruiz** [comment]
While I agree with Michael, I see value behind having a "single access
saga" using a marker interface (or equivalent), as even if it has scaling
limitations, it is also a pattern required sometimes. Unlimited scaling is
not always a requirement, sometimes a job (or a saga) may not be invoked so
often.

In such cases, having a 'simple' way of handling this 'single access sagas'
would be a win. Having to resort on a more complex code using an
intermediate saga in charge of serializing each job step, may be justified
just too much boilerplate for a simple issue. (IMHO)

**Comment by pruiz** [comment]
Hi,

I've been working on some generic approach to locking concurrent 'invocation' of sagas across multiple workers/processes. And as this may be an interesting addition to this discussion I've thought on sharing it here..

You can inspect the code at: https://github.com/evicertia/HermaFx/tree/pruiz/rebus-saga-locking/HermaFx.Rebus/Sagas, however, the code is against Rebus v1 (0.84.0), as this is what we are still using.

The locking implementation is provided by a simple "ISagaLockingManager' which at first I am implementing by using [redlock](https://github.com/samcook/RedLock.net), which is a a distributed locking algorithm using redis with auto-expiration and auto-recovery of dead clients.
But alternatively, an implementation using PostgreSQL advisory locks will follow.

This is still early and I have yet to figure out how to avoid a possible (not sure yet) deadlock involving cases where this is used in conjunction with a saga storage supporting updates to multiple sagas (like sql server)..

Also I am thinking on additional features like being able to specify enabling/disabling of locking on a saga basis, etc.

**Comment by mookid8000** [comment]
Sounds pretty cool!

You can avoid deadlocks if you acquire the lock in a specific order each time. So if it's possible, wait with acquiring the locks until you have all the instances, and then just acquire the locks ordered by the saga instances' IDs. It doesn't matter which order they're released in.

**Comment by pruiz** [comment]
Will all the handlers get instantiated before dispatching? Maybe on Rebus v2? If so, that could be a fine solution, yeah.. ;)

**Comment by mookid8000** [comment]
Yes, it is definitely possible with Rebus 2 :)

**Comment by mookid8000** [comment]
As an alternative to locking, I've added manual conflict resolution to Rebus 0.99.17. In a saga handler, you can override `ResolveConflict` and get the updated saga data as an argument, allowing you to e.g. merge its state into the instance you're currently updating.

Just need the build server to finish crunching, and then 0.99.17 will be out :)

**Comment by majic79** [comment]
I'd like to confirm that this occurs with FileSystem storage also (true, this is development only, however I sort of expect it to be a good starting point)

**Comment by mookid8000** [comment]
> I'd like to confirm that this occurs with FileSystem storage also (true, this is development only, however I sort of expect it to be a good starting point)

The `ResolveConflict` thing?

It can be invoked with all saga storages, as they all allow for 

1. loading a saga data instance
1. waiting an arbitrary duration of time, mutating the data in various ways
1. saving the saga data instance again

where (3) will either result in a `ConcurrencyException` (and subsequent rollback and retry) or the invocation of the `ResolveConflict` method.

None of the existing saga storages have any kind of pessimistic locking, they're optimists like we are 😉

**Comment by mookid8000** [comment]
one last thought though: maybe it would be neat to provide some kind of pessimistic locking mechanism whereby one could obtain locks on _anything_ – I am imagining an API that works like this:

    using(var myLock = await _locker.GetLock("arbitrary-lock-id"))
    {
         // do something that requires exclusive access to 
         // some resource identified by 'arbitrary-lock-id'
    }

which would asynchronously wait until the `arbitrary-lock-id` lock could be grabbed, and then release it again afterwards.

With this mechanism, it would be possible lock on different things, where a saga ID could be one thing, but it could also be another thing.

It's seldom a thing I need myself, but I often get asked how one can serialize access to one particular resource (e.g. a case no. or another entity ID thing) when e.g. running with multiple threads and high parallelism in an endpoint.

The easiest way to achieve this currently, is to have a parallelism of 1 – i.e. by serializing access to the world completely.

But if the locking thing was made, one could run with high parallelism but still serializing access on a per-resource-basis.

The locking thing could be implemented as an in-mem thing to begin with, which would be neat inside one single endpoint that could then increase its parallelism without constantly running into concurrency conflicts and subsequent rollback. But it could easily be extended to support distributed locking, implemented by SQL Server or whatever shared database is available.

**Comment by pruiz** [comment]
This sounds great, however, once you're inside a saga handler, trying to
lock/make-handler-wait-for-a-lock, will endup on a race condition between
two handlers for the same saga trying to persists an instance of SagaDate
with the same 'revision'. So only the first one will win, and the other one
still throw ConcurrencyEx.

That's because, both handlers, read the same 'revision' of SagaData, then
they serialized it's work/access by means of the lock, and once done, they
both tried to persist an updated SagaData with revision =
original-revision+1.. but only the first will be able to. :(

**Comment by mookid8000** [comment]
@pruiz yes, you're absolutely right that the grabbing of the lock would have to happen outside of the saga handler.... also, with the current implementation of `ISagaStorage` it would require loading the saga data twice, because the saga's ID would have to be known before the lock could be grabbed – and then, after having grabbed the lock, the saga data would have to be loaded again, because it could have been changed by someone else at that point – but at THAT point, we would know that we had gained exclusive access to the saga instance, and we could safely mutate & update it

**Comment by pruiz** [comment]
Yeah, it looks like this requires a pre-load-saga-step, where a lookup/find
of saga it's fired against db/store. And using the results of such lookup
(ie. SagaId) then try to obtain a lock, and if successfully obtained, you
can then continue down the path by actually loading the saga, invoking
handlers, etc..

**Comment by majic79** [comment]
Ah, no - I get concurrency exceptions using the file storage saga handler. I think a better solution is a message based one - sending saga updates on that and handling it in a single thread /reader

**Comment by jr01** [comment]
How about a centralized lock service/endpoint (conceptually similar to a central timeout service)? If you want a lock you send an ObtainLockMessage with a key to lock on and the message (+state) that you want back if you get the lock (i.e. MySagaHasTheLockMessage). When the lock is available, the lock service sends back that message to you. In the Handle(MySagaHasTheLockMessage) you do the exclusive work and reply a ReleaseLockMessage to the lock service to release the lock. 

Here is a proof-of-concept [gist](https://gist.github.com/jr01/e1505868d20f5785a4a66329b83d62dd) . I used a saga for implementing the lockservice. It should be easy to extend to support lock lease times, renew leases, semaphores (N parallel threads), etc.

**Comment by mookid8000** [comment]
@jr01 definitely an interesting experiment ... 

I think it would be better though if the lock could be grabbed in a synchronous fashion (of course still `await`ed, but synchronous in the same way that HTTP is synchronous), because this way the lock could be grabbed with a minimal overhead.

**Comment by MrMDavidson** [comment]
I've finally managed to get a chance to look at this again. Currently playing against the latest stable package in Nuget (3.1.5).

My current approach is an `IIncomingStep` implementation which is inserted prior to `LoadSagaDataStep`. This new step attempts to acquire a lock on the saga, if it can, then processing continues as per normal. If it cannot then the message is deferred. As long as we don't continue the current processing chain then the saga won't be loaded from the store and thus won't have a concurrent update. I haven't finished fleshing this out and making it nice but using the same repro case from above I'm able to successfully process the saga 100% of the time with more instances of the repro program and workers than my PC enjoys handling. The current rough implementation just uses a global semaphore as the lock (as it's all running on my machine). A more realistic implementation would use Redis / SQL Server app locks / whatever. The important thing is this would keep those implementation details out of the single access saga process.

```C#
public interface ISagaLockProvider {
    ISagaLock LockFor(object correlationId);
}

public interface ISagaLock : IDisposable {
    // Returns true if the lock was acquired, false otherwise
    Task<bool> AcquireAsync();

    // Release our lock
    Task  ReleaseAsync()
}

public class SagaLockStep : Rebus.Pipeline.IIncomingStep {
    private readonly ILog _log;
    private readonly IBus _bus;
    private readonly ISagaLockProvider _sagaLockProvider;

    public SagaLockStep(ILog log, IBus bus, ISagaLockProvider sagaLockProvider) {
        _log = log;
        _bus = bus;
        _sagaLockProvider = sagaLockProvider;
    }

    public async Task Process(IncomingStepContext context, Func<Task> next) {
	// Some of this code is pretty much copy/paste from the LoadSagaDataStep
        var helper = new SagaHelper();
        var sagaHandlers = context.Load<HandlerInvokers>()
            .Where(l => l.HasSaga)
            .ToList();

        if (sagaHandlers.Any() == false) {
            await next();
            return;
        }

        var message = context.Load<Message>();
        var body = message.Body;

        foreach (var invoker in sagaHandlers) {
            var props = helper.GetCorrelationProperties(body, invoker.Saga);
            var propsForMessage = props.ForMessage(body);

            foreach (var correlationProperty in propsForMessage) {
                var correlationId = correlationProperty.ValueFromMessage(body);

                using (ISagaLock sagaLock = await _sagaLockProvider.AcquireAsync(correlationId)) {
                    if (await sagaLock.AcquireAsync() == true) {
                        await next();
                        await sagaLock.ReleaseAsync();

			// Should we return here? I'm not sure if it's possible to have multiple sagas correlations on the one message? This might need to be refactored to attempt to lock all the sagas available (and gracefully defer and release all the locks if we can't) befeore running next()
                        return;
                    }

                    _log.Info($"Lock couldn't be acquired for ${correlationId} - deferring");
                    await _bus.Advanced.TransportMessage.Defer(TimeSpan.FromSeconds(10));
                    return;
                }
            }
        }

        await next();
    }
}
```

This is similar to your idea of having a global lock provider. I'm not up to date enough of the internals of Rebus to know if that's really required or a specific saga lock can be used. The other thing I'm not up to date on is the implementation of message deferral and how that's implemented; I notice that once a few deferrals occur the processing speed drops and I don't appear to be seeing the deferred messages in the queue. Are these, by default, held in memory?

**Comment by mookid8000** [comment]
> Are these, by default, held in memory?

Yes, they are 😱 

That will change in the future, though – in Rebus 4, all persistence options (subscriptions, sagas, and timeouts), which previously defaulted to in-mem implementations, will default to implementations that throw exceptions.

I have seen a couple of systems deployed to production unknowingly use the in-mem timeout manager, thus running a pretty big risk of losing deferred messages in the event of a restart. That has got to stop ✊ 

Super-elegant `SagaLockStep` btw – it should be trivial to replace the lock with a distributed version.

**Comment by mookid8000** [comment]
As an alternative to locking, remember that there's a `ResolveConflict` method that can be overridden on the `Saga<T>` class, enabling manual conflict resolution when it is possible to merge the states of the conflicting saga instances.

This also works well, even though it might be possibly to only sometimes merge the states, as it is allowed for the `ResolveConflict` method to simply throw a `ConcurrencyException` if it decides that it is not possibly to merge the states after all.

**Comment by MrMDavidson** [comment]
> Yes, they are 😱
> 
> That will change in the future, though – in Rebus 4, all persistence options (subscriptions, sagas, and timeouts), which previously defaulted to in-mem implementations, will default to implementations that throw exceptions.
> 
> I have seen a couple of systems deployed to production unknowingly use the in-mem timeout manager, thus running a pretty big risk of losing deferred messages in the event of a restart. That has got to stop ✊

Yup, that's fine, just thought I'd verify my assumptions. I'll double check this is still all good when using a persisted deferral implementation, but I think it should be.

I'll clean this up at work tomorrow and open a PR. Couple of questions so I can do that;
  1. What's the time frame for 4.x to be stable? Is it better to open it against 3.x (would by my preference so we can use it ASAP). 
  2. I notice you've started splitting everything up into separate nuget packages. I'd imagine that this would live in the base (unless you're against that...) and then the various packages (Eg. Rebus.SqlServer, Rebus.Azure or whatever) could provide implementations of the saga locking interface for distributed locks.
  3. For lack of intrusion on existing code I was thinking about making this only work if a handler has a saga marker. Eg. `public interface ISingleAccessSaga<TSagaDataType> : ISaga<TSagaDataType>` Do you have a preference here?
  4. Do you have any advice, re, the return statement / comments  in the example `SagaLockStep` I provided? If there's multiple handlers? Should I be acquiring a lock for all handlers of the saga before proceeding and if I can't deferring?

**Comment by MrMDavidson** [comment]
> As an alternative to locking, remember that there's a ResolveConflict method that can be overridden on the Saga<T> class, enabling manual conflict resolution when it is possible to merge the states of the conflicting saga instances.
> 
> This also works well, even though it might be possibly to only sometimes merge the states, as it is allowed for the ResolveConflict method to simply throw a ConcurrencyException if it decides that it is not possibly to merge the states after all.

I'll have a look at this tomorrow. I know with your original proposal for this it wasn't ideal for us because the concurrency conflict happens /after/ you've done work. Some of our processes (particularly the ones that we'd gain the most from moving to sagas!) have relatively long run times and are CPU intensive... so we'd really prefer not to do work and then end up discarding that work if we can't resolve the conflict. Those CPU minutes could have been used by another customer :)

**Comment by mookid8000** [comment]
> Some of our processes (particularly the ones that we'd gain the most from moving to sagas!) have relatively long run times and are CPU intensive

Ah... I usually avoid that by using the saga solely for orchestrating what should happen, delegating the actual work to request-/reply-based handlers – this usually makes the saga updates very fast, greatly reducing the risk of optimistic concurrency exceptions.

**Comment by MrMDavidson** [comment]
Re: Timeouts... I think there might be an issue here.... they don't seem to be doing what I'd expect. You can see in my snippet at the start of this issue Timeouts are configured to use SqlServer in the table `RebusTimeout`. But using both; `_bus.Defer()` and `_bus.Advanced.TransportMessage.Defer()` does not put the message into `RebusTimeout`. Infact compiling the `SqlServerTimeoutManager` into my test project and using that instead I can confirm that `SqlServerTimeoutManager::Defer()` is never called. Instead what I can see is happening is the message is put back into the queue but with a `visible` value set to whatever the deferred time is. This has the desired effect and survives restarts of workers. Is this how timeouts are meant to work?

**Comment by mookid8000** [comment]
oh, it's because the SQL transport obviates the need for timeout persistence – because SQL 😉 

you can see the "offending method" [here](https://github.com/rebus-org/Rebus.SqlServer/blob/master/Rebus.SqlServer/SqlServer/Transport/SqlServerTransport.cs#L349-L363)

**Comment by MrMDavidson** [comment]
Right - so the `config.Timeouts(cfg => cfg.StoreInSqlServer(...))` is more intended to be used if you're using some other transport? Got it.

**Comment by mario-d-s** [comment]
HI @mookid8000 , I came across this thread because we started having serious Saga concurrency issues in production since a couple of days.

The saga itself is the familiar pattern - send out batches of work and the same saga handles those batches, keeping progress/results in the saga data.

What started happening is that the number of instances that process messages was scaling up due to something unrelated. Then we started hitting those `ConcurrencyException`s, and looking through the source code myself I was shocked to see that **only 10 attempts** are made at optimistic concurrent updates...

I then found about https://github.com/rebus-org/Rebus/wiki/Saga-parallelism, and was happy to think there was a simple solution. I applied the config and... nothing. Exact same issue.

Any input from your side to counter this issue would be much appreciated.

If it's relevant at all - we use MySQL storage for both messages and sagas. Our transport is Amazon SQS.

**Comment by mookid8000** [comment]
For scenarios like this, it would be ideal if there was a distributed lock that could be used to eliminate all competition for each piece of saga data. For now though, I can make the number of conflict resolution attempts configurable. I've opened #944 to track that task.
